### PR TITLE
polyfill: durations accept fractional time values

### PR DESCRIPTION
**Pull request by ryzokuken**
Fixes: https://github.com/tc39/proposal-temporal/issues/938

/cc @justingrant @ptomato

### CODE REVIEW

File: polyfill/lib/regex.mjs
Comment (justingrant) [inline polyfill/lib/regex.mjs:36]: We should decide if 10+ fractional digits are OK or not, per comment above.

File: polyfill/test/duration.mjs
Comment (justingrant) [inline polyfill/test/duration.mjs:153]: IMHO, all these rounding tests should be done with both positive and negative durations, because a frequent cause of rounding errors in libraries is differences in handling negative vs. positive numbers.

**Review by justingrant — CHANGES_REQUESTED** [review]
Looks good, with one question and a few suggestions re: test cases.

File: polyfill/lib/ecmascript.mjs
Comment (ptomato) [inline polyfill/lib/ecmascript.mjs:921]: I'm not sure the spec language has first-class abstract operations like this, but, maybe consider instead splitting `BUILTIN_FIELDS` into `BUILTIN_DATETIME_FIELDS` and `DURATION_FIELDS`, and using `ToNumber` on duration fields?

**Review by ptomato — APPROVED** [review]
Looking good, the main change is that I think we should throw on more than our supported maximum number of decimal places, as per ISO 8601.

File: polyfill/lib/ecmascript.mjs
Comment (ryzokuken) [reply polyfill/lib/ecmascript.mjs:314]: > (Although then you do need to ensure that the seconds are not given in the string, if minutes are fractional.)

Exactly why I did it this way.

File: polyfill/lib/ecmascript.mjs
Comment (ryzokuken) [reply polyfill/lib/ecmascript.mjs:454]: done.

**Comment by ptomato** [comment]
I noticed from the CI output that the "order of operations" test262 tests supply non-integer input, those are breaking now instead of truncating it.

**Comment by littledan** [comment]
As I explained in https://github.com/tc39/proposal-temporal/issues/938#issuecomment-726896052 , I think this PR should be dropped.

**Comment by ptomato** [comment]
This PR still implements support for ISO strings such as `PT0.5H`, which is unrelated to the Number concern, I think?

File: spec/abstractops.html
Comment (ptomato) [inline spec/abstractops.html:102]:
```suggestion
        1. Set _minutes_ to the integral Number nearest to _mins_ in the direction of +0<sub>𝔽</sub>.
```
(also below)

(although I think you can also do more simply `Set _minutes_ to floor(_mins_).`)

File: spec/abstractops.html
Comment (ptomato) [inline spec/abstractops.html:990]: I think the grammar should show that the fraction is incompatible with the seconds part:
```
DurationMinutes :
    Digits MinutesDesignator DurationSeconds?
    Digits TimeFraction MinutesDesignator
```

File: spec/abstractops.html
Comment (ptomato) [inline spec/abstractops.html:1239]: DurationHoursFraction, DurationMinutesFraction, and DurationSecondsFraction need to be defined.

**Review by ptomato — APPROVED** [review]
I don't think we should block this PR on any of the remaining comments, since they are minor improvements, so I think punting them to a followup is fine.

File: spec/abstractops.html
Comment (ptomato) [reply spec/abstractops.html:102]: The linter was complaining about this line anyway so I'll push a new commit to fix it.

**Comment by ptomato** [comment]
OK, merging this now, we should solve the remaining comments in a follow up soon, ideally when the discussion in #938 is resolved.

File: spec/abstractops.html
Comment (ryzokuken) [reply spec/abstractops.html:989]: What we're doing right now is accept a duration with all the fractions, set the units to zero appropriately and handle them that way. We could change the spec text and instead of throwing at a later time, removing all those checks etc.

File: spec/abstractops.html
Comment (ryzokuken) [reply spec/abstractops.html:1238]: can we use `|TimeFractionalPart|` three times over?

File: spec/abstractops.html
Comment (ptomato) [reply spec/abstractops.html:989]: I am not 100% sure, but I think `PT0.5H30M` is actually illegal according to ISO 8601. In the polyfill it doesn't matter whether we throw because it's not accepted by the regex or we throw right after, but I think the spec grammar should match what's a legal string as much as possible.

File: spec/abstractops.html
Comment (ptomato) [reply spec/abstractops.html:1238]: Yes, I think you can just define them like `DurationHoursFraction : TimeFractionalPart` etc.

File: spec/abstractops.html
Comment (justingrant) [reply spec/abstractops.html:989]: Correct. Only the smallest unit is permitted to be fractional in ISO 8601.

Comment (ryzokuken) [reply spec/abstractops.html:989]: The string still throws, it just throws after the parsing stage (in `DurationHandleFractions`). I thought it was okay as long as we threw on the wrong format.

**Pull request merged**
